linera-core/src/client/received_log.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use linera_base::{crypto::ValidatorPublicKey, data_types::BlockHeight, identifiers::ChainId};
+use linera_chain::data_types::ChainAndHeight;
+
+/// Struct keeping track of the blocks sending messages to some chain, from chains identified by
+/// the keys in the map.
+pub(super) struct ReceivedLogs(
+    BTreeMap<ChainId, BTreeMap<BlockHeight, BTreeSet<ValidatorPublicKey>>>,
+);
+
+impl ReceivedLogs {
+    /// Creates a new instance of `ReceivedLogs`.
+    pub(super) fn new() -> Self {
```

Comment:
This doens't have to be `pub(super)`, it's used only down below in the `fold` as the starting point.

Maybe remove the method entirely and inline the `ReceivedLogs(BTreeMap::new())` there?